ambry-network/src/main/java/com.github.ambry.network/LocalNetworkClient.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.network;
+
+import com.github.ambry.clustermap.DataNodeId;
+import com.github.ambry.network.LocalRequestResponseChannel.LocalChannelRequest;
+import com.github.ambry.utils.ByteBufferInputStream;
+import com.github.ambry.utils.Time;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of {@link NetworkClient} that transmits requests and responses over a loopback network,
+ * where they can be dispatched directly to the AmbryRequests handler.
+ *
+ * This class is not thread safe.
+ */
+public class LocalNetworkClient implements NetworkClient {
+  private final LocalRequestResponseChannel channel;
+  private final NetworkMetrics networkMetrics;
+  private final Time time;
+  private boolean closed = false;
+  private final int processorId;
+  private static final Logger logger = LoggerFactory.getLogger(LocalNetworkClient.class);
+
+  /**
+   * Instantiates a LocalNetworkClient.
+   * @param channel the {@link RequestResponseChannel} to send requests to.
+   * @param processorId the id used to isolate requests and responses sent across the shared channel.
+   * @param networkMetrics the metrics to track the network related metrics
+   * @param time The Time instance to use.
+   */
+  public LocalNetworkClient(LocalRequestResponseChannel channel, int processorId, NetworkMetrics networkMetrics,
+      Time time) {
+    this.channel = channel;
+    this.processorId = processorId;
```

Comment:
I still don't quite understand the `processorId`, what the difference between processorId and correlationId? Looks like the processorId is associated with LocalNetworkClient. So, there are several LocalNetworkClients sharing same channel?

Reply:
Yes, that is because each OperationController creates its own NetworkClient sharing the same channel, and when any of them retrieves its responses, it needs to get only the ones for the requests that it sent.  But it's not quite the same as correlationId since it doesn't correlate request with response.  It just segregates all the responses into separate buckets based on which client sent them.  I coped the name from the existing network client/channel classes.  I can change it to clientId if you prefer.